lib/mix/lib/mix/tasks/release.ex

```diff
@@ -0,0 +1,913 @@
+    File.mkdir_p!(bin_path)
+
+    for os <- Keyword.get(release.options, :include_executables_for, [:unix, :windows]) do
+      [{start, contents} | clis] = cli_for(os, release)
+      start_path = Path.join(bin_path, start)
+
+      unless File.exists?(start_path) do
+        File.write!(start_path, contents)
+        executable!(start_path)
+      end
+
+      for {filename, contents} <- clis do
+        path = Path.join(bin_path, filename)
+        File.write!(path, contents)
+        executable!(path)
+      end
+
+      unless File.exists?(elixir_bin_path) do
+        Mix.raise("Could not find bin files from Elixir installation")
+      end
+
+      for {filename, contents} <- elixir_cli_for(os, elixir_bin_path, release) do
+        path = Path.join(release.version_path, filename)
+        File.write!(path, contents)
+        executable!(path)
+      end
+    end
+  end
+
+  # TODO: Implement windows CLI
+  defp cli_for(_os, release) do
+    [
+      {"start", start_template(name: release.name)},
+      {"#{release.name}", cli_template(name: release.name)}
+    ]
+  end
+
+  defp elixir_cli_for(:unix, bin_path, release) do
+    [
+      {"elixir",
+       Path.join(bin_path, "elixir")
+       |> File.read!()
+       |> String.replace(~s[ -pa "$SCRIPT_PATH"/../lib/*/ebin], "")
+       |> replace_erts_bin(release, ~s["$SCRIPT_PATH"/../../erts-#{release.erts_version}/bin/])},
+      {"iex", File.read!(Path.join(bin_path, "iex"))}
+    ]
+  end
+
+  defp elixir_cli_for(:windows, bin_path, release) do
+    [
+      {"elixir.bat",
+       Path.join(bin_path, "elixir.bat")
+       |> File.read!()
+       |> String.replace(~s[goto expand_erl_libs], ~s[goto run])
+       |> replace_erts_bin(release, ~s["%dp0\\..\\..\\erts-#{release.erts_version}\\bin\\"])},
+      {"iex.bat", File.read!(Path.join(bin_path, "iex"))}
+    ]
+  end
+
+  defp executable!(path), do: File.chmod!(path, 0o744)
+
+  defp replace_erts_bin(contents, release, new_path) do
+    if release.erts_source do
+      String.replace(contents, ~s[ERTS_BIN=""], ~s[ERTS_BIN=#{new_path}])
+    else
+      contents
+    end
+  end
+
+  ## Templates
+
+  embed_text(:vm_args, ~S"""
+  ## Do not load code from filesystem as all modules are preloaded
+  -mode embedded
+
+  ## Disable the heartbeat system to automatically restart the VM
+  ## if it dies or becomes unresponsive. Useful only in daemon mode.
+  ##-heart
+
+  ## Number of diry schedulers doing IO work (file, sockets, etc)
+  ##+SDio 5
+
+  ## Increase number of concurrent ports/sockets
+  ##-env ERL_MAX_PORTS 4096
+
+  ## Tweak GC to run more often
+  ##-env ERL_FULLSWEEP_AFTER 10
+  """)
+
+  embed_template(:start, ~S"""
+  #!/bin/sh
+  set -e
+  # Feel free to edit this file in anyway you want
+  # To start your system using IEx: . $(dirname "$0")/<%= @name %> start iex
+  # To start it as a daemon using IEx: . $(dirname "$0")/<%= @name %> daemon iex
+  . $(dirname "$0")/<%= @name %> start
```

Comment:
Will `<%= @name %>` be a valid unix path?

Reply:
I will validate this. Good catch.